### PR TITLE
Auto updating local data if too old

### LINKED ISSUE #81: Feature request: auto-update tldr pages
Every two weeks, I get the message 
`Local data is older than two weeks, use --update to update it.`

I'd like TLDR to just auto-update when it needs to. 

I think this could be easily/simply implemented by giving users the option to set an environment variable, e.g. `TLDR_AUTOUPDATE`. If this variable is set, we could have `check_localdate` automatically call `update_localdb` if necessary. None of this would impact users who do not set the environment variable.

I'd be happy to work on it and submit a PR if people think it's an OK idea.

### PR DESCRIPTION
**Pull request by riesentoaster**
## What does it do?
If the local data is too old, it now no longer asks you to manually update it but attempts to automatically do so. Regardless of whether it succeeds, the output is shown.

## Why the change?
It is annoying to manually update it. See also #81.

## How can this be tested?
Change the date in `~/.tldrc/date` to something further back than the past two weeks, run any `tldr` command like `tldr tldr`

## Where to start code review?
Very few lines changed in `local.c`

## Relevant tickets?
Closes #81.

## Questions?
This is just a first attempt. If you'd like me to change how this is implemented, I will gladly do that. I just wanted to start the conversation.

### CODE REVIEW

**Comment by riesentoaster** [comment]
Would one of the maintainers care to take a look at this? Pinging @MasterOdin, @owenvoke, whoever else wants do.

**Comment by kbdharun** [comment]
> Once again pinging <> maintainer. Any feedback on this?

Hi, I too think this will be useful, unfortunately, I am not proficient with C even though I know it as bit so I can't help with this. I hope this issue will be addressed soon.

**Review by kbdharun — APPROVED** [review]

**Review by SethFalco — COMMENTED** [review]
Change makes sense to me, but just some UX (color) concerns!

File: src/local.c
Comment (SethFalco) [inline src/local.c:60]: This has output which ends up with the bold red, even though it probably shouldn't be.
Could we reset the foreground color and disable bold before this `if`?

Alternatively, we may want to consider switching from red (looks like an error) to a yellow color to more visually it's a warning/heads-up, and nothing actually went wrong.

Without introducing any additional colors, I think this would be better:

```c
if (difftime > (60 * 60 * 24 * 7 * 2)) {
    /* *INDENT-OFF* */
    fprintf(stdout, "Local data is older than two weeks, attempting to automatically update it...\n");
    if (update_localdb(0)) {
        fprintf(stdout, "%s%s", ANSI_BOLD_ON, ANSI_COLOR_CODE_FG);
        fprintf(stdout, "Failed to automatically update local data.\n");
        fprintf(stdout, "%s%s", ANSI_COLOR_RESET_FG, ANSI_BOLD_OFF);
    }
    fprintf(stdout, "\n");
    /* *INDENT-ON* */
}
```
> PS: I'm not much of a C person, is there any reason why multiple ANSI codes weren't done like that before? :thinking: 

### Initial UX (unchanged)

image
> When a first does their first command and no local DB exists.

### Current behavior for outdated DB

image

### PR behavior for outdated DB

image
> Red tends to indicate an error or in the case before, that action was required. But if this is no longer the case, this should change.

### Proposed behavior for outdated DB

image
> Colors/width now aligns with the initial user experience.

File: src/local.c
Comment (SethFalco) [inline src/local.c:59]:
```suggestion
        fprintf(stdout, "Local data is older than two weeks, attempting to update it...\n");
```

No need for the word automatic here, the user knows it's automatic because it was automatically done for them!

File: src/local.c
Comment (SethFalco) [inline src/local.c:61]:
```suggestion
            fprintf(stdout, "Failed to update local data.\n");
```

**Review by SethFalco — COMMENTED** [review]
Actually, also want to nitpick at the copy.

No need for the word automatic here, the user knows it's automatic because it was automatically done for them!

It's also a bit jarring to refer to it as `data` in one string, and then `database` immediately after. ^-^'

It's referred to as `local database` explicitly everywhere else, let's stick with that!

image
> See how the only instances of `local data` (without having `base` affixed to it) are the instances from this PR.

**Review by SethFalco — CHANGES_REQUESTED** [review]
I'm so sorry for the review spam, I just keep getting new thoughts when I think I'm finished! I think this will be a no-go as it is now. 

There are users who specifically don't want it to automatically try to update the cache, i.e. due to poor network conditions.

* See https://github.com/tldr-pages/tldr-c-client/issues/27

Having logic to preserve the original behavior as default, and only perform the update when toggled, i.e. via environment variable or client setting should be a prerequisite.

> I think this could be easily/simply implemented by giving users the option to set an environment variable, e.g. `TLDR_AUTOUPDATE`.
>
> — https://github.com/tldr-pages/tldr-c-client/issues/81

**Comment by riesentoaster** [comment]
Thank you so much for looking at it!

I agree with both the color and wording points, those seem reasonable. For your last point: That is a bit of an issue, can't have it both ways. The current archive is \~6MB. 

My personal preference would be to automatically download the updates and having an option with either an argument and/or an environment variable to disable it. If I am currently in an environment where the network is slow enough for me to care about it, I'd probably like one of the following behaviours:

1. If the automatic download takes more than say 3 seconds, it is cancelled, and the output is displayed anyway along with a message that tells the user to manually update the database.
2. The download is started and a message is displayed that tells the user what exact command they'd have to use to prevent the automatic update, so when they get annoyed, they could just kill the process (`ctrl-c`) and copy paste the line to instantly get the output.
3. The old version is displayed automatically and the download is only started afterwards. If there was an update to the explanation of this specific command, the new version is printed (and potentially the old one deleted). This way, the user immediately gets an answer and can continue working while the download still is in progress. If they want to use the same terminal immediately, they can either kill the process or move it to the background.

I only ever invoke `tldr` manually and I'm not sure when you'd ever do so in a script, so having a manual cancellation option seems ok to me.

Opinions on this?

**Comment by SethFalco** [comment]
> My personal preference would be to automatically download the updates and having an option with either an argument and/or an environment variable to disable it.

Overall, this is actually my preference as well. If this were a new client, I'd expect to see this.

However, when a project is already out and being used in the wild, I think it's often best to make changes like these iteratively. So to preserve the original behavior and make the option to opt-in for automatic updates, and at a later time change the default if it's deemed safe to do so. However, I'm happy to concede if other maintainers think it's not a big deal.

> If the automatic download takes more than say 3 seconds, it is cancelled, and the output is displayed anyway along with a message that tells the user to manually update the database.

This wouldn't be good UX. No one likes inconsistent behavior. An implementation like this leads to unpredictable output since it entirely depends on network conditions. Even on a fast network, sometimes it can be congested.

> The download is started and a message is displayed that tells the user what exact command they'd have to use to prevent the automatic update

This is viable. :+1: 

> The old version is displayed automatically and the download is only started afterwards. If there was an update to the explanation of this specific command, the new version is printed (and potentially the old one deleted).

This wouldn't be a good UX. It'd be like the terminal equivalent of CLS (Content Layout Shift) in browsers. At that point, I think it'd be better to just make users wait for the update. Imo this is more work, for a lesser experience.

---

From my perspective, so long as it's configurable (opt-in or opt-out), I'd be willing to approve the PR. But if the default behavior changes, I'll wait for another maintainer to approve/merge just to solidify if we're happy to change it.

**Comment by riesentoaster** [comment]
Thank you for your feedback! I'm just pitching ideas and as long as we end up agreeing on something, I am happy.

So, assuming we implement a `--prevent-update` flag: What do we do with the `--update` flag?
- Removing it would introduce more breaking changes (but one could also argue that since we're already touching it, might as well…).
- Not removing it leads to a weird situation when both flags are used.
  - Do we just print an error message?
  - Or give one priority and print an error message as well?
  - Or just give one priority and ignore the other?
- Otherwise I'd say that using no flags would just trigger the auto update.
- Using just the `--prevent-update` flag would output the page from the local database. (Would we even check for the age and print a message if it is too old then? Since the user specifically asked not to update it?)
- Using just the `--update` flag would have the same behaviour as it has now: Update the database and don't print any other output.

**Comment by SethFalco** [comment]
Hmm, I don't think a flag would be ideal for this. It's not like users will use `--prevent-update` every single time they use the client. ^-^'

Removing the `--update` arg isn't an option, all clients with a cache support `-u` / `--update`, this is also defined in the [client specifications](https://github.com/tldr-pages/tldr/blob/main/CLIENT-SPECIFICATION.md#arguments). Even if it's redundant, it should stay.

Assuming the default behavior changes to auto-update, users can just set an environment variable like `TLDR_AUTO_UPDATE_DISABLED=1` if they don't want the cache to update automatically.

Think of it like [`DOTNET_CLI_TELEMETRY_OPTOUT`](https://learn.microsoft.com/en-us/dotnet/core/tools/telemetry#how-to-opt-out) or [`NUXT_TELEMETRY_DISABLED`](https://github.com/nuxt/telemetry#opting-out).

This is basically the inverse of the original proposal to have an environment variable to enable auto-updates.

If the environment variable is defined, auto-updates are not performed unless the `--update` flag is present. For such users, when the local database is > 2 weeks old, or a command doesn't exist, it will not update unless `--update` is provided.

**Comment by riesentoaster** [comment]
That seems like an elegant solution. I implemented it, added a comment to the message explaining that the auto update is about to happen and one to the README for documentation.

I also added a check that prevents `tldr` from doing its online lookup in case a requested page isn't available locally, as you suggested.

With those changes this PR would be ready for a new round of reviews in my opinion, so if anyone would like to take a look, please do!

File: src/local.c
Comment (SethFalco) [inline src/local.c:55]: Could you address the indentation/formatting, please?

File: src/parser.c
Comment (SethFalco) [inline src/parser.c:212]: Due to the new flow, this should have `return 0;` after it, and line 215 would have `return 1;`. Welcome to correct me if you think I'm wrong.

If auto updates are enabled, and we find new pages, we should return 0 because everything was successful.

However, if auto-updates are disabled or after updating the page was not found, we should return 1 to indicate we failed to get a page. (With your PR as-is, we return a success status but offer no output to users when the page doesn't exist.)

It may also be worth adding a print in tldr.c with how the user can check for updates, for example:

```
This page doesn't exist yet!
Submit new pages here: https://github.com/tldr-pages/tldr
Automatic updates are disabled, you could try updating the local database manually with: tldr --update
```
> Note in this case I'd expect the 3rd message to appear in red, and only if `TLDR_AUTO_UPDATE_DISABLED` it set.

Reference: https://github.com/tldr-pages/tldr-c-client/blob/main/src/tldr.c#L178

I'm partial to the output I proposed, so feedback from anyone is welcome.

**Review by SethFalco — CHANGES_REQUESTED** [review]
Overall, this looks pretty good to me, thanks for taking the time to make implement this. Once the issues I noted are resolved, I'd be happy to approve this.

We'll need other maintainers to review the UX regarding prints specific for disabled automatic updates, but I'll request more reviews once everything else is fine. :thinking:

File: src/parser.c
Comment (riesentoaster) [reply src/parser.c:212]: Agreed on the return value. I'd propose a slight change in the wording of the additional error message:

```
$ tldr non_existing_command       
This page doesn't exist yet!
Submit new pages here: https://github.com/tldr-pages/tldr
Checking the online database was skipped because automatic updates are disabled.
You could try updating the local database manually with: tldr --update
```

If any of those lines should be printed in color, I'd argue it should be the first one, since it's the one actually saying what the error was. The last two are a hint at best but since the env variable has to be set manually by the user, I don't think this should be considered an error.

**Review by SethFalco — CHANGES_REQUESTED** [review]
Almost there, one petty change and I think this is good.
I hope other maintainers agree. ^-^

**Review by SethFalco — APPROVED** [review]
I think this is a good change. :+1: 
Thank you for your contribution!

I'm happy to approve, but I would prefer to wait for an OG maintainer to chime and confirm they are happy with the default behavior changing before we merge.

I'll pester people in Matrix if no one wants to come around to review this soon.

Edit: I think this also solves #27, so after this, I'll also close that and post a comment with how to achieve the desired behavior.

**Review by kbdharun — APPROVED** [review]
LGTM, Thanks for your contribution.

**Comment by riesentoaster** [comment]
> I'll pester people in Matrix if no one wants to come around to review this soon.

@SethFalco Any progress on this?

**Review by MasterOdin — APPROVED** [review]
Thanks for your contribution. I think it may make sense to release this as part of a major release to better signify the change in behavior around when updates get downloaded.

**Comment by kbdharun** [comment]
> Thanks for your contribution. I think it may make sense to release this as part of a major release to better signify the change in behavior around when updates get downloaded.

Offtopic: I think we create one for our Python and Elixr client too, as we updated the tldr source from `master` to `main` (+ we were planning to deprecate the `master` branch next month)

**Pull request merged**
